text/2024-01-27-autohinting.md

```diff
@@ -0,0 +1,266 @@
+outlines are requested from the font. This document will focus on the latter.
+
+> Note: the techniques discussed here are implemented in 
+[skrifa](https://github.com/googlefonts/fontations/tree/2c9212d99184498327aa44b2069f8183025f5fa9/skrifa/src/outline/autohint)
+but are based heavily on the original work done by David Turner, Werner
+Lemberg and others in the FreeType
+[autofit](https://gitlab.freedesktop.org/freetype/freetype/-/tree/f92c96550ad763639158587974cf11067ace743d/src/autofit)
+module.
+
+> Secondary note: this is a tour of the autohinting process. The actual
+implementation is fairly complex and not amenable to a full description here.
+Links have been provided to the relevant pieces of code so that the
+interested reader can follow along. Improvements to this document are
+welcome!
+
+Let's talk autohinting:
+
+* [The stylish blues](#the-stylish-blues)- style classification and metrics
+computation
+* [The importance of connection](#the-importance-of-connection)- topological
+analysis of outlines
+* [Time for a change](#time-for-a-change)- making adjustments to an outline
+
+## The stylish blues
+
+Before we can starting mucking about with outlines, we need to gather some
+information from the font to determine exactly *how* the outlines should be
+adjusted.
+
+This information falls into two categories:
+
+### Style classification ([src](https://github.com/googlefonts/fontations/blob/6eff17a15ca0caaea27eb130fa959d529d58625b/skrifa/src/outline/autohint/style.rs))
+
+Styles are effectively groups of glyphs that should be hinted in the same way
+with the same set of metrics (see the next section). These groups are broadly
+defined by Unicode script but may be further divided by specific OpenType
+features when the glyphs produced by those features would have different sizes
+and/or positions (e.g. subscripts, superscripts, small capitals).
+
+Style classification proceeds in three phases:
+1. Enumerate the character to glyph id mapping table and assign styles
+ roughly based on Unicode script. For example, the glyph id associated
+ with the character "A" would be assigned the "Latin default" style.
+2. Process the layout subsitution tables for known script/feature combinations
+ and assign the appropriate styles to glyphs that are *output* from those
+ substitutions. For example, any glyph id that is generated by the `smcp`
+ feature under the `latn` script would be assigned the "Latin small caps"
+ style.
+ 3. Process the layout substitution tables again, for each script, and assign
+ the default style for that script to any remaining unassigned glyph ids that
+ appear as outputs. This captures things like ligatures.
+
+This process is done for all glyphs at once when the autohinter state is
+initialized. The structure of a font file (lacking efficient mappings from
+glyph id to character or to layout feature) makes it cost prohibitive to do
+style classification for a single glyph.
```

Comment:
This seems to imply potential for a font format enhancement, suggest noting this or why that's not the case